File: src/core/report.ts
import { EnergyViolation } from '../types';
import { SeverityCounts, emptyCounts } from './severityCounts';

// invariant: this module must not import fs/child_process/vscode — it only aggregates and
// renders data the caller already has, so it stays testable with hand-built fixtures
// (see src/test/integration/report.test.ts) and reusable from both the CLI's scan and diff modes
//
// decision: split into report.ts (aggregate scoring + markdown table), reportDiff.ts (base/head
// diffing), and reportHuman.ts (the human-readable risk report) — the combined file had grown to
// 16 functions across three largely independent domains, which is exactly the sprawl this tool's
// own coherence detector flags. This file re-exports the other two so existing imports of
// './report' keep working unchanged.

export interface FileResult {
    filePath: string;
    violations: EnergyViolation[];
}

export type { SeverityCounts };

export interface FileSummary {
    filePath: string;
    score: number;
    counts: SeverityCounts;
    byType: Record<string, number>;
}

export interface AggregateSummary {
    files: FileSummary[];
    totalScore: number;
    totalCounts: SeverityCounts;
}

// decision: weights (1/4/9) make high-severity violations dominate the score without parsing
// numeric complexity out of detector message strings — see src/core/detectors/cyclomatic.ts and
// cognitive.ts, which only embed their numeric complexity in `message`, not as a structured field.
// invariant: these weights are load-bearing for diff continuity (diffSummaries below) — changing
// them changes the meaning of a previously reported delta, so treat this as a versioned constant.
const SEVERITY_WEIGHT: SeverityCounts = { low: 1, medium: 4, high: 9 };

export function summarizeFile(result: FileResult): FileSummary {
    const counts = emptyCounts();
    const byType: Record<string, number> = {};

    for (const violation of result.violations) {
        counts[violation.severity] += 1;
        byType[violation.type] = (byType[violation.type] ?? 0) + 1;
    }

    const score =
        counts.low * SEVERITY_WEIGHT.low + counts.medium * SEVERITY_WEIGHT.medium + counts.high * SEVERITY_WEIGHT.high;

    return { filePath: result.filePath, score, counts, byType };
}

export function summarize(results: FileResult[]): AggregateSummary {
    const files = results.map(summarizeFile);
    const totalCounts = emptyCounts();
    let totalScore = 0;

    for (const file of files) {
        totalScore += file.score;
        totalCounts.low += file.counts.low;
        totalCounts.medium += file.counts.medium;
        totalCounts.high += file.counts.high;
    }

    return { files, totalScore, totalCounts };
}

export function hasBlockingViolations(counts: SeverityCounts): boolean {
    return counts.medium > 0 || counts.high > 0;
}

export function renderMarkdownReport(summary: AggregateSummary): string {
    const cleanCount = summary.files.filter((f) => f.score === 0).length;
    const lines: string[] = [];

    lines.push('# Energy State Report');
    lines.push('');
    lines.push(
        `**${summary.files.length} file${summary.files.length === 1 ? '' : 's'} scanned** — ${cleanCount} clean, ${summary.files.length - cleanCount} with violations`
    );
    lines.push('');
    lines.push('| File | Score | High | Medium | Low |');
    lines.push('| --- | --- | --- | --- | --- |');

    for (const file of summary.files) {
        lines.push(
            `| ${file.filePath} | ${file.score} | ${file.counts.high} | ${file.counts.medium} | ${file.counts.low} |`
        );
    }

    lines.push('');
    lines.push(
        `**Total score: ${summary.totalScore}** (${summary.totalCounts.high} high, ${summary.totalCounts.medium} medium, ${summary.totalCounts.low} low)`
    );

    return lines.join('\n');
}

export * from './reportDiff';
export * from './reportHuman';

File: src/Core/Report.fs
module Energy.Core.Report

open Energy.Core.Violation

type SeverityCounts = { Low: int; Medium: int; High: int }

type FileResult =
    { FilePath: string
      Violations: EnergyViolation list }

type FileSummary =
    { FilePath: string
      Score: int
      Counts: SeverityCounts
      ByType: Map<string, int> }

type AggregateSummary =
    { Files: FileSummary list
      TotalScore: int
      TotalCounts: SeverityCounts }

let emptyCounts = { Low = 0; Medium = 0; High = 0 }

let private addViolation counts violation =
    match violation.Severity with
    | Low -> { counts with Low = counts.Low + 1 }
    | Medium ->
        { counts with
            Medium = counts.Medium + 1 }
    | High -> { counts with High = counts.High + 1 }

let private score counts =
    counts.Low + 4 * counts.Medium + 9 * counts.High

// invariant: weights 1/4/9 are the published report/diff continuity metric; only structured
// severity contributes, never numbers embedded in detector messages.
let summarizeFile result =
    let counts = result.Violations |> List.fold addViolation emptyCounts

    let byType =
        result.Violations
        |> List.fold
            (fun types violation ->
                let name = violationTypeName violation.Type
                Map.change name (Option.defaultValue 0 >> (+) 1 >> Some) types)
            Map.empty

    { FilePath = result.FilePath
      Score = score counts
      Counts = counts
      ByType = byType }

let summarize results =
    let files = results |> List.map summarizeFile

    let totalCounts =
        files
        |> List.fold
            (fun counts file ->
                { Low = counts.Low + file.Counts.Low
                  Medium = counts.Medium + file.Counts.Medium
                  High = counts.High + file.Counts.High })
            emptyCounts

    { Files = files
      TotalScore = files |> List.sumBy _.Score
      TotalCounts = totalCounts }

let hasBlockingViolations counts = counts.Medium > 0 || counts.High > 0

let renderMarkdownReport summary =
    let cleanCount =
        summary.Files |> List.filter (fun file -> file.Score = 0) |> List.length

    let fileCount = summary.Files.Length
    let suffix = if fileCount = 1 then "" else "s"

    let rows =
        summary.Files
        |> List.map (fun file ->
            sprintf
                "| %s | %d | %d | %d | %d |"
                file.FilePath
                file.Score
                file.Counts.High
                file.Counts.Medium
                file.Counts.Low)

    [ "# Energy State Report"
      ""
      sprintf
          "**%d file%s scanned** — %d clean, %d with violations"
          fileCount
          suffix
          cleanCount
          (fileCount - cleanCount)
      ""
      "| File | Score | High | Medium | Low |"
      "| --- | --- | --- | --- | --- |" ]
    @ rows
    @ [ ""
        sprintf
            "**Total score: %d** (%d high, %d medium, %d low)"
            summary.TotalScore
            summary.TotalCounts.High
            summary.TotalCounts.Medium
            summary.TotalCounts.Low ]
    |> String.concat "\n"
